File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.cpp
#include<string>
#include <deque>
#include <iostream>

using namespace std;

deque<char> st;

bool solution(string s)
{
    bool answer = true;

    int len=s.length();
    
    for(int i=0;i<len;i++){
        char& c=s.at(i);
        
        if(c=='('){
            st.push_back(c);
        }
        else if(c==')'){
            if(st.empty()){
                answer=false;
                break;
            }
            else{
                if(st.back()=='('){
                    st.pop_back();
                }
                else{
                    answer=false;
                    break;
                }
            }
        }
        else{
            answer=false;
            break;
        }
    }
    
    if(!st.empty()){
        answer=false;
    }

    return answer;
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;
class Solution {
    boolean solution(String s) {
        boolean answer = true;

        // [실행] 버튼을 누르면 출력 값을 볼 수 있습니다.
        System.out.println("Hello Java");
        
        Deque<Character> dq=new ArrayDeque<>();
        char[] str=s.toCharArray();
        for(char c:str){
            if(c=='('){
                dq.addLast(c);
            }
            else{
                if(dq.isEmpty()){
                    return false;
                }
                if(dq.pollLast()!='('){
                    return false;
                }
            }
        }
        
        if(!dq.isEmpty()){
            return false;
        }

        return true;
    }
}